src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	/* ensure that the task is not NULL */
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
 	if (commandType != CMD_SELECT)
 	{
-		bool topLevel = true;
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock() && xactParticipantHash == NULL)
+		{
+			MemoryContext oldContext = NULL;
+			ListCell *placementCell = NULL;
+
+			xactParticipantHash = CreateXactParticipantHash();
+
+			oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+			foreach(placementCell, task->taskPlacementList)
+			{
+				ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+				XactParticipantKey participantKey;
+				XactParticipantEntry *participantEntry = NULL;
+				bool entryFound = false;
+
+				PGconn *connection = NULL;
+				uint64 *shardIdPtr = NULL;
+
+				MemSet(&participantKey, 0, sizeof(participantKey));
+				strncpy(participantKey.nodeName, placement->nodeName, MAX_NODE_LENGTH);
+				participantKey.nodePort = placement->nodePort;
+
+				participantEntry = hash_search(xactParticipantHash, &participantKey,
+											   HASH_ENTER, &entryFound);
+				Assert(!entryFound);
+
+				connection = GetOrEstablishConnection(placement->nodeName,
+													  placement->nodePort);
+				if (connection != NULL)
+				{
+					PGresult *result = PQexec(connection, "BEGIN");
+					if (PQresultStatus(result) != PGRES_COMMAND_OK)
+					{
+						WarnRemoteError(connection, result);
+						PurgeConnection(connection);
+
+						connection = NULL;
+					}
+
+					PQclear(result);
+				}
+
+				participantEntry->connection = connection;
+
+				shardIdPtr = AllocateUint64(placement->shardId);
+
+				participantEntry->shardIds = list_make1(shardIdPtr);
+			}
+
+			MemoryContextSwitchTo(oldContext);
+		}
```

Comment:
:grin: It was, but during the rewrite, this logic got shorter, so I again inlined it. I'll un-inline it (again).